src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java

```diff
@@ -0,0 +1,211 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.worker;
+
+import com.google.devtools.build.lib.shell.Subprocess;
+import com.google.devtools.build.lib.shell.SubprocessBuilder;
+import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse;
+import com.google.devtools.build.lib.vfs.Path;
+import java.io.File;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+
+/** An intermediate worker that sends request and receives response from the worker processes. */
+public class WorkerMultiplexer extends Thread {
+  /**
+   * WorkerMultiplexer is running as a thread on its own. When worker process
+   * returns the WorkResponse, it is stored in this map and wait for
+   * WorkerProxy to retrieve the response.
+   */
+  private Map<Integer, InputStream> workerProcessResponse;
+  /**
+   * A semaphore to protect workerProcessResponse object.
+   */
+  private Semaphore semWorkerProcessResponse;
+  /**
+   * After sending the WorkRequest, WorkerProxy will wait on a semaphore to be
+   * released. WorkerMultiplexer is responsible to release the corresponding
+   * semaphore in order to signal WorkerProxy that the WorkerResponse has been
+   * received.
+   */
+  private Map<Integer, Semaphore> responseChecker;
+  /**
+   * A semaphore to protect responseChecker object.
+   */
+  private Semaphore semResponseChecker;
+  /**
+   * The worker process that this WorkerMultiplexer should be talking to.
+   */
+  private Subprocess process;
+
+  private Thread shutdownHook;
+  private Integer workerHash;
+
+  WorkerMultiplexer(Integer workerHash) {
+    semWorkerProcessResponse = new Semaphore(1);
+    semResponseChecker = new Semaphore(1);
+    responseChecker = new HashMap<>();
+    workerProcessResponse = new HashMap<>();
+    this.workerHash = workerHash;
+
+    final WorkerMultiplexer self = this;
+    this.shutdownHook =
```

Comment:
This shutdown hook also needs to be removed otherwise `destroyMultiplexer` is called twice:

 1. From this shutdown hook
 2. From `WorkerProxy`'s shutdown hook (`WorkerProxy.destroy` -> `WorkerMultiplexerManager.removeInstance(workerHash);` -> `multiplexerInstance.get(workerHash).destroyMultiplexer();`)

Reply:
Good catch, thanks.